File: app/models/stack_box.py
import enum
from datetime import datetime

from sqlalchemy import BigInteger, DateTime, Enum, ForeignKey, Integer, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column

from app.database import Base


class StackBoxType(str, enum.Enum):
    folder = "folder"
    page = "page"
    canvas = "canvas"
    edgeless = "edgeless"


class StackBox(Base):
    __tablename__ = "stack_boxes"

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True)
    workspace_id: Mapped[int] = mapped_column(
        BigInteger, ForeignKey("workspaces.id", ondelete="CASCADE"), nullable=False
    )
    parent_id: Mapped[int | None] = mapped_column(
        BigInteger, ForeignKey("stack_boxes.id", ondelete="CASCADE"), nullable=True
    )
    type: Mapped[StackBoxType] = mapped_column(
        Enum(StackBoxType, name="stack_box_type"), nullable=False, default=StackBoxType.folder
    )
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    description: Mapped[str] = mapped_column(Text, nullable=False, default="")
    icon: Mapped[str | None] = mapped_column(String(64), nullable=True)
    cover_url: Mapped[str | None] = mapped_column(String(512), nullable=True)
    sort_order: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    created_by: Mapped[int | None] = mapped_column(
        BigInteger, ForeignKey("users.id", ondelete="SET NULL"), nullable=True
    )
    updated_by: Mapped[int | None] = mapped_column(
        BigInteger, ForeignKey("users.id", ondelete="SET NULL"), nullable=True
    )
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())

File: lib/stackbox/stack_boxes/stack_box.ex
defmodule Stackbox.StackBoxes.StackBox do
  use Ecto.Schema
  import Ecto.Changeset

  schema "stack_boxes" do
    field :type, Ecto.Enum, values: [:folder, :page, :canvas, :edgeless], default: :folder
    field :name, :string
    field :description, :string, default: ""
    field :icon, :string
    field :cover_url, :string
    field :sort_order, :integer, default: 0

    belongs_to :workspace, Stackbox.Workspaces.Workspace
    belongs_to :parent, __MODULE__

    field :created_by, :id

    belongs_to :created_by_user, Stackbox.Accounts.User,
      foreign_key: :created_by,
      define_field: false

    field :updated_by, :id

    belongs_to :updated_by_user, Stackbox.Accounts.User,
      foreign_key: :updated_by,
      define_field: false

    field :created_at, :utc_datetime, read_after_writes: true
    field :updated_at, :utc_datetime, read_after_writes: true
  end

  @doc """
  Creation changeset. `created_by`/`updated_by` are stamped server-side from
  the authenticated user (not client-settable), mirroring
  `backend/app/routers/stack_boxes.py`'s `create_stack_box`.
  """
  def create_changeset(stack_box, attrs, creator_id) do
    stack_box
    |> cast(attrs, [
      :workspace_id,
      :parent_id,
      :type,
      :name,
      :description,
      :icon,
      :cover_url,
      :sort_order
    ])
    |> validate_required([:workspace_id, :type, :name])
    |> validate_length(:name, max: 255)
    |> put_change(:created_by, creator_id)
    |> put_change(:updated_by, creator_id)
    |> foreign_key_constraint(:workspace_id)
    |> foreign_key_constraint(:parent_id)
  end

  @doc """
  Update changeset. Excludes `workspace_id` (no re-parenting to another
  workspace) and stamps `updated_by` server-side.
  """
  def update_changeset(stack_box, attrs, updater_id) do
    stack_box
    |> cast(attrs, [:parent_id, :type, :name, :description, :icon, :cover_url, :sort_order])
    |> validate_length(:name, max: 255)
    |> put_change(:updated_by, updater_id)
    |> foreign_key_constraint(:parent_id)
  end
end
